flink-formats/flink-csv/src/main/java/org/apache/flink/formats/csv/CsvRowDataDeserializationSchema.java

```diff
@@ -0,0 +1,458 @@
+			.map(RowType.RowField::getType)
+			.map(this::createNullableConverter)
+			.toArray(DeserializationRuntimeConverter[]::new);
+		final String[] fieldNames = rowType.getFieldNames().toArray(new String[0]);
+		final int arity = fieldNames.length;
+
+		return jsonNode -> {
+			int nodeSize = jsonNode.size();
+
+			validateArity(arity, nodeSize, ignoreParseErrors);
+
+			GenericRowData row = new GenericRowData(arity);
+			for (int i = 0; i < arity; i++) {
+				JsonNode field;
+				// Jackson only supports mapping by name in the first level
+				if (isTopLevel) {
+					field = jsonNode.get(fieldNames[i]);
+				} else {
+					field = jsonNode.get(i);
+				}
+				if (field == null) {
+					row.setField(i, null);
+				} else {
+					row.setField(i, fieldConverters[i].convert(field));
+				}
+			}
+			return row;
+		};
+	}
+
+	/**
+	 * Creates a runtime converter which is null safe.
+	 */
+	private DeserializationRuntimeConverter createNullableConverter(LogicalType type) {
+		final DeserializationRuntimeConverter converter = createConverter(type);
+		return jsonNode -> {
+			if (jsonNode == null || jsonNode.isNull()) {
+				return null;
+			}
+			try {
+				return converter.convert(jsonNode);
+			} catch (Throwable t) {
+				if (!ignoreParseErrors) {
+					throw t;
+				}
+				return null;
+			}
+		};
+	}
+
+	/**
+	 * Creates a runtime converter which assuming input object is not null.
+	 */
+	private DeserializationRuntimeConverter createConverter(LogicalType type) {
+		switch (type.getTypeRoot()) {
+			case NULL:
+				return jsonNode -> null;
+			case BOOLEAN:
+				return this::convertToBoolean;
+			case TINYINT:
+				return jsonNode -> Byte.parseByte(jsonNode.asText().trim());
+			case SMALLINT:
+				return jsonNode -> Short.parseShort(jsonNode.asText().trim());
+			case INTEGER:
+			case INTERVAL_YEAR_MONTH:
+				return this::convertToInt;
+			case BIGINT:
+			case INTERVAL_DAY_TIME:
+				return this::convertToLong;
+			case DATE:
+				return this::convertToDate;
+			case TIME_WITHOUT_TIME_ZONE:
+				return this::convertToTime;
+			case TIMESTAMP_WITH_TIME_ZONE:
+			case TIMESTAMP_WITHOUT_TIME_ZONE:
+				return this::convertToTimestamp;
+			case FLOAT:
+				return this::convertToFloat;
+			case DOUBLE:
+				return this::convertToDouble;
+			case CHAR:
+			case VARCHAR:
+				return this::convertToString;
+			case BINARY:
+			case VARBINARY:
+				return this::convertToBytes;
+			case DECIMAL:
+				return createDecimalConverter((DecimalType) type);
+			case ARRAY:
+				return createArrayConverter((ArrayType) type);
+			case ROW:
+				return createRowConverter((RowType) type, false);
+			case MAP:
+			case MULTISET:
+			case RAW:
+			default:
+				throw new UnsupportedOperationException("Unsupported type: " + type);
+		}
+	}
+
+	private boolean convertToBoolean(JsonNode jsonNode) {
+		if (jsonNode.isBoolean()) {
+			// avoid redundant toString and parseBoolean, for better performance
```

Comment:
Why we can't use `node.asBoolean()` directly?

Reply:
We want to support this case `{"f0": "true"}` where the value is a string node, not boolean node. This is also how `CsvRowDeserializationSchema` process boolean and numeric types. I just add a fast path (avoid toString and parse if it's a boolean node).

Reply:
`TextNode` also has a correct implementation for `asBoolean`.

Reply:
The problem of `asBoolean`, `asLong` is that they eat the parse exception and fall back to the default value (0 for Long, and false for Boolean). But we want to keep the parse excepton to let users be aware of it, and can use `ignore-parse-error` option to turn it off.